rskj-core/src/main/java/org/ethereum/core/BlockTxSignatureCache.java

```diff
@@ -21,53 +21,51 @@
 
 import co.rsk.core.RskAddress;
 import co.rsk.remasc.RemascTransaction;
-import co.rsk.util.MaxSizeHashMap;
-
 
 public class BlockTxSignatureCache extends SignatureCache {
 
     private static final int MAX_CACHE_SIZE = 900;
 
-    private SignatureCache internalCache;
+    private final SignatureCache internalCache;
 
     public BlockTxSignatureCache(SignatureCache internalCache) {
+        super(MAX_CACHE_SIZE, false);
         this.internalCache = internalCache;
-        addressesCache = new MaxSizeHashMap<>(MAX_CACHE_SIZE,false);
     }
 
     @Override
-    public RskAddress getSender(Transaction transaction) {
+    public synchronized RskAddress getSender(Transaction transaction) {
```

Comment:
If I'm getting it correctly, this synchronization is not required. After removing the manual initialization of `addressesCache`, both it and `internalCache` are initialised with `Collections.synchronizedMap` and you are not iterating over a "collection-converted" version of the map.

Am I missing a scenario when this is required maybe?

Reply:
I guess `Collections.synchronizedMap()` would only make those two collections synchronised, meaning their operations, like `put()` or `remove()`, would be synchronised and thread-safe, but I believe the idea here is to make the entire `getSender` method synchronised, so no race conditions will occur in there

Reply:
Yes, what you mention about the collections is correct.

It is the race condition what I am not able to spot, or at least a problematic one, but most probably I'm failing to find it :)

The only thing I can imagine is the time between `addressesCache.get` and `addressesCache.put`, where two simultaneous threads could consider the address as new and both would compute the sender. Not sure what is better in this case, this extra computation or the synchronization penalty, but I guess this was measured or considered.

Reply:
Yup, that was the idea. Maybe we can get rid of this modifier on `getSender` for `ReceivedTxSignatureCache` since there are no "cache write" inside it so a race condition would not occur, but in the rest of the cases (`storeSender` method on `ReceivedTxSignatureCache` and `storeSender` and `getSender` methods on `BlockTxSignatureCache`) should be protected against race conditions, since both methods do write to the cache. Maybe the worst case scenario is not that bad, but better to avoid it.

Reply:
those collections are instances the `MaxSizeHashMap` class, which has a side affect in the `put` method - it removes older items if the cache is full.

For instance, [this](https://github.com/rsksmart/rskj/blob/IRIS-3.1.0/rskj-core/src/main/java/org/ethereum/core/BlockTxSignatureCache.java#L50-L54) sort of code would be problematic in multi-threaded environments, as `addressesCache.put()` may delete an item in one thread, so `internalCache.getSender(transaction)` would return `null` in some other one, which obviously breaks the logic.

Also by having `SignatureCache` class synchronised at its own level, I think we would not need to make those two collections synchronised.

In general, I personally prefer to have such kind of "safeguards" in place, especially when behaviour methods or its dependencies may change later.

Reply:
I'd also rewrite [this](https://github.com/rsksmart/rskj/blob/IRIS-3.1.0/rskj-core/src/main/java/org/ethereum/core/BlockTxSignatureCache.java#L50-L54) kind of blocks - we could only call `internalCache.getSender()` and check its result for `null` ( without calling `internalCache.containsTx()` )

Reply:
I see your point in the [previous message](https://github.com/rsksmart/rskj/pull/1687#discussion_r783722101) @Vovchyk 👍

Reply:
I've made an additional refactor regarding [this](https://github.com/rsksmart/rskj/pull/1687#discussion_r783728513).